server/rest/src/test/java/org/infinispan/rest/security/AuthorizationTest.java

```diff
@@ -0,0 +1,166 @@
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.rest.assertion.ResponseAssertion;
+import org.infinispan.rest.authentication.impl.BasicAuthenticator;
+import org.infinispan.rest.helper.RestServerHelper;
+import org.infinispan.security.AuthorizationPermission;
+import org.infinispan.security.Security;
+import org.infinispan.security.impl.IdentityRoleMapper;
+import org.infinispan.server.core.security.simple.SimpleServerAuthenticationProvider;
+import org.infinispan.test.AbstractInfinispanTest;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.TestCacheManagerFactory;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "rest.security.AuthorizationTest")
+public class AuthorizationTest extends AbstractInfinispanTest {
+
+   public static final String REALM = "realm";
+   private HttpClient client;
+   private RestServerHelper restServer;
+
+   static final Subject ADMIN;
+   static final Map<AuthorizationPermission, Subject> SUBJECTS;
+
+   static {
+      // Initialize one subject per permission
+      SUBJECTS = new HashMap<>(AuthorizationPermission.values().length);
+      for (AuthorizationPermission perm : AuthorizationPermission.values()) {
+         SUBJECTS.put(perm, TestingUtil.makeSubject(perm.toString() + "_user", perm.toString()));
+      }
+      ADMIN = SUBJECTS.get(AuthorizationPermission.ALL);
+   }
+
+   @BeforeSuite
+   public void beforeSuite() throws Exception {
+      client = new HttpClient();
+      client.start();
+   }
+
+   @AfterSuite
+   public void afterSuite() throws Exception {
+      Security.doAs(ADMIN, (PrivilegedAction<Void>) () -> { restServer.stop(); return null; });
+      client.stop();
+   }
+
+   @AfterMethod
+   public void afterMethod() {
+      Security.doAs(ADMIN, (PrivilegedAction<Void>) () -> { restServer.clear(); return null; });
+      if (restServer != null) {
+         Security.doAs(ADMIN, (PrivilegedAction<Void>) () -> { restServer.stop(); return null; });
+      }
+   }
+
+   @Test
+   public void testPermissions() throws Exception {
+      //given
+      initRestServer();
+
+      // when reader puts
+      ContentResponse put = request("default", "key", "reader", "password", HttpMethod.PUT)
+            .content(new StringContentProvider("value"))
+            .send();
+
+      //then
+      ResponseAssertion.assertThat(put).isForbidden();
+
+      //when writer puts
+      put = request("default", "key", "writer", "password", HttpMethod.PUT)
+            .content(new StringContentProvider("value"))
+            .send();
+
+      //then
+      ResponseAssertion.assertThat(put).isOk();
+
+      // when writer gets
+      ContentResponse get = request("default", "key", "writer", "password", HttpMethod.GET).send();
+
+      //then
+      ResponseAssertion.assertThat(get).isForbidden();
```

Comment:
Could you please move those test cases into separate methods?

Reply:
Done